### ISSUE TITLE
Limitations on customizing the install target

### ISSUE THREAD

**Issue by rib** [mesonbuild/meson#2998]
Hi,

We have a project that doesn't fit the mould of wanting to install everything under a conventional `prefix` + `libdir` + `bindir` + `datadir` ... where we don't want our subprojects assuming they are installing to a conventional Linux package prefix.

In our case we're building a Unity3d plugin (usually cross-compiled for Android) where we want the `install` target (at least during development) to copy files across to a Unity project.

The master project itself is dependent on multiple subprojects which have declared that they want to install various libraries and headers, but those decisions on what to install don't really make sense when embedded in another project (at least not in this case, but I do wonder how often it would ever make sense).

Should target install decisions really be inherited? It seems to me like projects shouldn't make assumptions about what parent projects they might be embedded in and shouldn't impose decisions about installation.

Since it doesn't currently work for us to use `add_install_script` while we can't practically block the subprojects from installing files (without modifying the `meson.build` for upstream projects we otherwise don't modify) then we tried to use `run_target('install_plugin', [install_script.py])` instead. The problem now though is that `MESONINTROSPECT` (which our install script depends on) is not set in the environment so we can't easily be sure to run the same version of meson to enumerate the built targets.

I'm not sure of the best solution here.

Would it be acceptable to ensure `MESONINTROSPECT` is set for `run_target()` scripts?
Would it be acceptable to provide some way of telling meson to not install anything from subprojects?

**Comment by rib** [comment]
Just to mention a workaround I'm using for now:

Since `run_command()` will have `MESONINTROSPECT` set, I can use that to store the value in a meson variable:
```
python3 = import('python3').find_python()
introspect_cmd = run_command(python3, '-c',
    'import os; print(os.getenv("MESONINTROSPECT", ""))').stdout().strip()
```
And then I pass `introspect_cmd` as an argument to `run_target('install_plugin', [ 'install_script.py', '--meson-introspect', introspect_cmd, ...])`

**Comment by jpakkane** [comment]
> Would it be acceptable to ensure MESONINTROSPECT is set for run_target() scripts?

This should already be the case. Run targets are invoked via the commandrunner and it sets that among a bunch of other envvars.

**Comment by rib** [comment]
hmm, right, I just double checked, and `MESONINTROSPECT` is set as you say. I guess I must have misinterpreted a different error the other day that led me to believe it wasn't there - sorry.

I still wish there was some way to really control the `install` target instead of having to make up an alternative. Just today I've mistyped `ninja install` several times instead of `ninja install_plugin` and installed a load of spurious files I don't want.

**Comment by jpakkane** [comment]
A very hacky solution is to install stuff to a temp dir, say `/tmp/foo` and then have a custom script to copy things from there to where you really want them to be.

**Comment by rib** [comment]
Yeah, from the original discussion above it looks like I made a mistake claiming it wasn't set. (Note this is an _old_ issue from 2018 so sorry that I'm not aware if anything has changed recently)

I think overall installing stuff is quite a separate problem from building stuff. On Linux there are lots of standard conventions for laying out files for libraries and packages and it makes some sense that Meson has default install behaviour that can work for some common cases but considering more general use cases I think it would be good to just provide an escape hatch that let's you basically override what install does for a top level project and easy to parse meta data.

In our case we were fine with being able to write a python script that would refer to the json introspection data to find the various build outputs and arrange them how we needed, it just would have been convenient if it would have integrated with `meson install` and `ninja install`.
